nvflare/tool/code_pre_installer/README.md

```diff
@@ -0,0 +1,255 @@
+
+## Usage
+
+### Command Line Interface
+
+```bash
+nvflare pre-install -a /path/to/application.zip -p /opt/nvflare/apps -s site-1 [-ts /local/custom] [-debug]
+```
+
+Arguments:
+- `-a, --application`: Path to application code zip file (required)
+- `-p, --install-prefix`: Installation prefix (default: /opt/nvflare/apps)
+- `-s, --site-name`: Target site name e.g., site-1, server (required)
+- `-ts, --target_shared_dir`: Target shared directory path (default: /local/custom)
+- `-debug, --debug`: Enable debug mode
+
+### Example
+
+```bash
+# Install application code for site-1
+nvflare pre-install -a /path/to/myapp.zip -s site-1
+
+# Install with custom paths
+nvflare pre-install -a /path/to/myapp.zip -p /custom/install/path -s site-1 -ts /custom/shared/path
+
+# Install with debug output
+nvflare pre-install -a /path/to/myapp.zip -s site-1 -debug
+```
+
+## Application Code Structure
+
+The application zip file should have the following structure:
+
+```
+application/
+├── app_name/
+│   ├── meta.json
+│   ├── app_site-1/
+│   │   └── custom/
+│   │       └── site_specific_code.py
+│   └── app_site-2/
+│       └── custom/
+│           └── site_specific_code.py
+└── application-share/
+    └── shared_code.py
+```
+
+- `app_name/`: Application directory containing site-specific code
+- `meta.json`: Application metadata file
+- `app_site-*/custom/`: Site-specific custom code directories
+- `application-share/`: Shared code directory
+
+## Installation Paths
+
+- Application code: `<install-prefix>/<app-name>/`
+- Shared resources: `/local/custom/`
+
+## Error Handling
+
+The installer will fail if:
+- Job structure zip is invalid or missing required directories
+- meta.json is missing or invalid
+- Site directory not found and no default apps available
+- Installation directories cannot be created
+- File operations fail
+- Package installation fails (if requirements.txt present)
+
+## Notes
+
+- Existing files may be overwritten
+- Python path is automatically configured for shared packages
+- All file permissions are preserved during installation
+- Network access needed if requirements.txt present
+- Can use private PyPI server by configuring pip
+- The tool will extract site-specific code to the installation prefix
+- Shared code will be installed to the target shared directory
+- The application zip file will be cleaned up after installation
+- Installation paths must be writable by the current user
+
+## Using Pre-installed Code
+
+### In Job Configuration (JSON)
+```json
+{
+    "task_script": "${NVFLARE_INSTALL_PREFIX}/src/client.py",
+    "other_config": "..."
+}
+```
+
+### In Development
+#### JSON Config
+```bash
+export NVFLARE_INSTALL_PREFIX=""  # Empty for development
+```
+
+#### Python Code
+```python
+task_script_path = "src/client.py"
+```
+
+### In Production
+#### JSON Config
+```bash
+export NVFLARE_INSTALL_PREFIX="/opt/nvflare/jobs/fedavg"  # Set for production
```

Comment:
Should be "export NVFLARE_INSTALL_PREFIX="/opt/nvflare/jobs/"?
since the "fedavg" is actually coming from each job's meta?

Reply:
what happen when there are multiple jobs?

```